ArduCopter/landing_gear.cpp

```diff
@@ -12,14 +12,30 @@ void Copter::landinggear_update()
     // last status (deployed or retracted) used to check for changes, initialised to startup state of landing gear
     static bool last_deploy_status = landinggear.deployed();
 
-    // if we are doing an automatic landing procedure, force the landing gear to deploy.
-    // To-Do: should we pause the auto-land procedure to give time for gear to come down?
-    if (control_mode == LAND ||
-       (control_mode == RTL && flightmode_rtl.landing_gear_should_be_deployed()) ||
-       (control_mode == AUTO && flightmode_auto.landing_gear_should_be_deployed())) {
+    // gear sw warned used to warn the operator once of the gear switch being in the wrong position
+    static bool gcs_gear_sw_warned = false;
+
+    // If we've entered an flight mode which requries the gear to auto-deploy, deploy and lock the gear
+    if (landinggear_deploy_modes()) {
         landinggear.set_position(AP_LandingGear::LandingGear_Deploy_And_Keep_Deployed);
     }
 
+    // If we've exited an auto-deploy mode, check the landing gear switch.
```

Comment:
transition logic between flight modes should be hooked in the flightmode exit function.